File: go-backend-interview-prep/05_errors_and_testing/01_error_basics/main.go
package main

import (
	"errors"
	"fmt"
	"strconv"
)

// divide returns an error when the divisor is zero.
func divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, errors.New("division by zero")
	}
	return a / b, nil
}

// parseAndDouble parses a string to int, doubles it.
func parseAndDouble(s string) (int, error) {
	n, err := strconv.Atoi(s)
	if err != nil {
		return 0, fmt.Errorf("parseAndDouble(%q): %v", s, err)
	}
	return n * 2, nil
}

func main() {
	// --- Example 1: basic error check ---
	result, err := divide(10, 3)
	if err != nil {
		fmt.Println("ERROR:", err)
	} else {
		fmt.Printf("10 / 3 = %.2f\n", result)
	}

	// division by zero
	result, err = divide(10, 0)
	if err != nil {
		fmt.Println("ERROR:", err)
	} else {
		fmt.Printf("10 / 0 = %.2f\n", result)
	}

	// --- Example 2: wrapping context ---
	val, err := parseAndDouble("42")
	if err != nil {
		fmt.Println("ERROR:", err)
	} else {
		fmt.Println("parseAndDouble(\"42\") =", val)
	}

	val, err = parseAndDouble("abc")
	if err != nil {
		fmt.Println("ERROR:", err)
	} else {
		fmt.Println("parseAndDouble(\"abc\") =", val)
	}

	// --- Example 3: creating errors ---
	e1 := errors.New("something went wrong")
	e2 := fmt.Errorf("failed to load config: %v", e1)
	fmt.Println(e2)
}

File: go-backend-interview-prep/05_errors_and_testing/01_error_basics/main.py
"""Error basics -- Python equivalent of the Go example."""


def divide(a: float, b: float) -> float:
    """Divide a by b. Raises ValueError on division by zero."""
    if b == 0:
        raise ValueError("division by zero")
    return a / b


def parse_and_double(s: str) -> int:
    """Parse string to int and double it."""
    try:
        n = int(s)
    except ValueError as e:
        raise ValueError(f"parse_and_double({s!r}): {e}") from e
    return n * 2


def main() -> None:
    # --- Example 1: basic error check ---
    try:
        result = divide(10, 3)
        print(f"10 / 3 = {result:.2f}")
    except ValueError as e:
        print("ERROR:", e)

    try:
        result = divide(10, 0)
        print(f"10 / 0 = {result:.2f}")
    except ValueError as e:
        print("ERROR:", e)

    # --- Example 2: wrapping context ---
    try:
        val = parse_and_double("42")
        print(f'parse_and_double("42") = {val}')
    except ValueError as e:
        print("ERROR:", e)

    try:
        val = parse_and_double("abc")
        print(f'parse_and_double("abc") = {val}')
    except ValueError as e:
        print("ERROR:", e)

    # --- Example 3: creating errors ---
    e1 = ValueError("something went wrong")
    e2 = RuntimeError(f"failed to load config: {e1}")
    print(e2)


if __name__ == "__main__":
    main()
